module-bsp/board/rt1051/puretx/bsp/keyboard/keyboard.cpp

```diff
@@ -67,24 +68,47 @@ namespace bsp::keyboard
                 }
             }
 
-            gpio->EnableInterrupt(1U << static_cast<uint32_t>(BoardDefinitions::KEYBOARD_RF_BUTTON));
+            gpio->EnableInterrupt(1U << static_cast<std::uint32_t>(BoardDefinitions::KEYBOARD_RF_BUTTON));
         }
+    } // anonymous namespace
 
-        void clearAllIRQs()
-        {
-            std::uint8_t clearIRQs = 0xFF;
-            i2cAddr.subAddress     = REG_INT_STAT;
-            i2c->Write(i2cAddr, reinterpret_cast<std::uint8_t *>(&clearIRQs), 1);
+    bool write(const drivers::I2CAddress &addr, const std::uint8_t *txBuff, const std::size_t size)
+    {
+        for (auto i = 0; i < maxTransmissionRetry; ++i) {
+            auto ret = i2c->Write(addr, txBuff, size);
```

Comment:
Is this call blocking i.e. we're waiting here for the end of transmission?

Let's consider a case when transmission failed. Does the write start just right after the previous one? If that's the case and the error was caused by some condition of the i2c could this condition be still present? If that's the case maybe we should wait a while between tries?

Reply:
Yes, we are waiting for the end of transmission.

In this case, next transmission starts immediately after the previous one. So that's a good tip. We can wait a bit and try to write/read again. How long should we wait 1 or 5ms?

Reply:
Your guess is as good as mine. :)
The times you propose looks reasonable.

I'd only add that we may consider waiting different ammounts of time for each call in case this condition was time-related. And because it may be caused by another part in the system sending messages (which is typically done in constant intervals) maybe these time values should be something more random? E.g. 1253us and 4723us?

Reply:
I suppose that we can increase delay time in case when the operation fails eg. the next delay time will be two times longer.  Do you want to random delay every time or just hardcoded delays? I think that this is a bit overcomplicated. :)

Reply:
I was thinking about something hardcoded. Increasing the time is ok too to play well with `maxTransmissionRetry`.

> I think that this is a bit overcomplicated.

Yes, proabably. It was just a question so if you think this is not needed then I'm ok with it. :)

Reply:
These are all valid points, but I want you to think about responsibility here for a second. Do you think it is a good idea to have this simple method perform retransmission logic? Imho it is not ;). The context here is too narrow to perform any meaningful action(why if in specific context 3 retransmission will be not enough?). Why not return a specific error code and then let the higher layers do something with it?